### DISCUSSION TITLE
Writing a boost-histogram to a root file created in uproot (with proper sum-of-weights errors)

### DISCUSSION THREAD

**Discussion by ellkay7** [scikit-hep/boost-histogram#787] [Q&A]
I am copying a question I posted on stackoverflow a while ago, since this seems like a better place to ask: https://stackoverflow.com/questions/65951054/writing-a-boost-histogram-to-a-root-file-created-in-uproot

I am currently trying to read in a ROOT tree with uproot, book and fill histograms with boost-histogram and write these back out to a new ROOT file. This seems to work very well for the bin contents, but I can't seem to get the correct bin errors for the histograms in the final ROOT file. When I read these histograms, `h.GetBinError(b)` seems to be equal to `h.GetBinContent(b)`, while I would like to have the familiar sumW2 errors which are normally produced for a ROOT histogram. My workflow is as follows, using python 3.8:

```
import boost_histogram as bh
import uproot3 as uproot 
bins = np.array( [1,2,3,4], dtype='d' )
hist = bh.Histogram( bh.axis.Variable(bins))
```
I then iterate over the input tree using uproot4.iterate, to produce an awkward array for each 'chunk', denoted here by e:

```
hist.fill(e["pt"], weight=(e["weight"]))
with uproot.recreate(outfilename) as outfile:
   outfile["hist_pt"] = hist.to_numpy()
```
I am not sure how to get the sumw2 weights in the construction of the boost-histogram, but based on the documentation I changed my histogram definition to:

`hist = bh.Histogram( bh.axis.Variable(bins), storage=bh.storage.Weight() )`
However, when I run my code, this causes issues with the `.to_numpy()` line. The error says: `what = uproot3_methods.convert.towriteable(what)" "TypeError: cannot perform reduce with flexible type`

Is there some other way to propagate sumW2-style errors from a boost-histogram to a ROOT file created in uproot? Or perhaps there is an error in my understanding of this storage= variable.... Any help would be greatly appreciated.

Perhaps my question is now outdated... is there some better way to efficiently convert data from an awkward array in uproot, which can then be saved as a histogram to a root file?

**Comment by andrzejnovak** [comment]
Hi @special-kay-0, the key issue here is calling `.to_numpy()` which without any qualifiers converts to a tuple of `(values, bins)` or such. Try the following example I had at hand (using `bh` instead of `hist` should work just as well)

```
import uproot
import hist
import numpy as np
import mplhep as hep

h = hist.new.Reg(20, 0, 20, name='msd').Weight().fill(np.random.normal(10, 6, 1000))

fout = uproot.recreate('test.root')
fout['test'] = h
fout.close()

fin = uproot.open('test.root')
h_read = fin['test']
fin.close()

```

**Comment by henryiii** [comment]
Why are you using uproot 3? Uproot 4 will natively take a boost-histogram or Hist object (as in the other answer).
